### PR TITLE
common,osd: set osd_memory_target based on POD_MEMORY_REQUEST or cgroup limit

### PR DESCRIPTION
**Pull request by liewegas**
This is an initial, minimal pass at this.

- default osd_memory_target to .8x the cgroup limit, if present
- set osd_memory_target to POD_MEMORY_REQUEST, if present

There are many possible enhancements:

- we could make the .8 tunable
- we could make the .8 dynamically tunable at runtime

I'm most interested in first getting something basic in place for kubernetes, and second for ceph in a random container (say, ceph-nano or ceph-ansible).  I think this covers both when combined with @leseb's rook change at https://github.com/rook/rook/pull/2764

### CODE REVIEW

**Comment by liewegas** [comment]
@travisn @leseb first question/confirmation: for the minimal thing, do we need to look at POD_MEMORY_LIMIT?  IIRC it's possible we get POD_MEMORY_LIMIT but no POD_MEMORY_REQUEST.  And even if that's possible with kube, perhaps rook can be sure to set the request value on the pod.

**Comment by liewegas** [comment]
I've updated to do both: default to .8x the cgroup limit, and also pay attention to the POD_MEMORY_REQUEST value if it is present.  WDYT?

File: src/common/config.cc [hunk 1]
```diff
@@ -474,6 +475,30 @@ void md_config_t::parse_env(ConfigValues& values,
...
+	       CONF_ENV, nullptr);
+      break;
+    }
+  }
+  uint64_t cgroup_limit = 0;
+  if (get_cgroup_memory_limit(&cgroup_limit) == 0) {
+    // this is an "environment" property in the general sense, so
+    // populate the option at level CEPH_DEFAULT.  that way a user can
+    // override with the normal (e.g., mon) config.
+    cgroup_limit *= .8;
```
Comment (batrick) [inline src/common/config.cc]: I'd suggest a config option for this `.8`.

Comment (liewegas) [reply src/common/config.cc]: Yeah... the problem is that snowballs pretty quickly.  That config option might come from the mon, or a config file, or via admin socket 'config set', which means we don't have the value now. That means we need to move this logic somewhere much later.  Let me think about where that would be.

Comment (liewegas) [reply src/common/config.cc]: Ok, I think teh way to do this is downstream of the core config parsing.  Take a look...

File: src/common/config.cc [hunk 2]
```diff
@@ -474,6 +475,16 @@ void md_config_t::parse_env(ConfigValues& values,
       _set_val(values, tracker, dir, *o, CONF_ENV, &err);
     }
   }
+  const char *pod_req = getenv("POD_MEMORY_REQUEST");
+  if (pod_req) {
```
Comment (leseb) [inline src/common/config.cc]: Note that `POD_MEMORY_REQUEST` will be zero if not specified in the pod definition. During pod's scheduling, if `memory.request` is not set it equals `memory.limit`, however, that's not reflected in pod's properties when running.

Comment (liewegas) [reply src/common/config.cc]: yeah, but (1) we don't want to set osd_memory_target to the actual limit or we'll quickly crash, and (2) we pick up that limit * the ratio from the cgroup directly in the next patch

Comment (liewegas) [reply src/common/config.cc]: oh, you mean we shoudl do something different if the value is zero?

Comment (leseb) [reply src/common/config.cc]: Yeah, if value is `0` then we switch to reading `/sys`. So I guess the condition should be `if (pod_req) && (pod_req != 0)`.

File: src/common/options.cc [hunk 3]
```diff
@@ -4013,6 +4013,11 @@ std::vector<Option> get_global_options() {
     .add_see_also("bluestore_cache_autotune")
     .set_description("When tcmalloc and cache autotuning is enabled, try to keep this many bytes mapped in memory."),
 
+    Option("osd_memory_target_cgroup_limit_ratio", Option::TYPE_FLOAT, Option::LEVEL_ADVANCED)
+    .set_default(0.8)
```
Comment (batrick) [inline src/common/options.cc]: `    .set_min_max(0.0, 1.0)`

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -4195,6 +4202,23 @@ void BlueStore::_set_blob_size()
 
 int BlueStore::_set_cache_sizes()
 {
+  // set osd_memory_target *default* based on cgroup limit?
+  // (do this before we fetch the osd_memory_target value!)
+  double cgroup_ratio = cct->_conf.get_val<double>(
+    "osd_memory_target_cgroup_limit_ratio");
+  if (cgroup_ratio) {
```
Comment (batrick) [inline src/os/bluestore/BlueStore.cc]: Converting a `double` to `bool` makes me squeamish. How about `cgroup_ratio > 0.0`?

File: src/common/options.cc [hunk 5]
```diff
@@ -4013,6 +4013,11 @@ std::vector<Option> get_global_options() {
...
+    .add_see_also("osd_memory_target")
+    .set_description("Set the default value for osd_memory_target to the cgroup memory limit (if set) times this value"),
```
Comment (batrick) [inline src/common/options.cc]: Should also mention that setting `0.0` disables changing the `osd_memory_target` based on the cgroup memory limit. Maybe in the long description?

File: src/common/config.cc (hunk 2 above)
Comment (salanki) [reply src/common/config.cc]: Is it possible that the check for `pod_req != 0` was never implemented? I am seeing some strange behavior where the memory target is set to 100G.

Comment (leseb) [reply src/common/config.cc]: @salanki I had the same concern, I checked with @liewegas and it seems that if the value is 0 then it's treated as if it isn't set at all. So we are covered.

**Pull request merged**